### ISSUE TITLE
How to import haunted — in the future!

### ISSUE THREAD

**Issue by chase-moskal** [matthewp/haunted#98]
hello friends, i absolutely love the way the javascript ecosystem is evolving, and there are a lot of emerging ideas — so please pitch in with your thoughts about these best practices and changes to haunted that i'm proposing

this proposal calls for a major version bump, as it involves some important changes to the way haunted is packaged and imported

## let's imagine using haunted, but in the future 🕶️

in the future, the coolest way to include haunted (or anything really) in your app, will be:

**direct imports in the browser, empowered by [import maps](https://github.com/WICG/import-maps)**
1. configure an import map to resolve `haunted` and `lit-html` from unpkg or anywhere else
2. `import {component} from "haunted"`, and get cracking! it's that easy!
    - haunted and any other libraries will use the `lit-html` version you specify in your import map
    - everything can work together in harmony :beers: 

**bundling**  
also in the future, we'll be able to run that same import map through a bundler to prepare a nice compact bundle for our app, to save on http-requests in production

**we can use this pattern today**  
the best part is that we can use this pattern today via [es-module-shims](https://github.com/guybedford/es-module-shims)!  
another alternative is unpkg's really cool `?module` feature.  
well, we can't use it with haunted *just yet* — only after we all agree to and implement the changes in this proposal :)

**what's the difference compared to how we import haunted right now?**  
if you're using a traditional bundler, like with npm, i think there'll be no change for you.  
otherwise for es module importers, you'll have to tell haunted which version of lit-html to use, and where to find it.  
if you don't, the browser will throw an error.  
this is already how it works for other packages like lit-element.  
*"isn't that just annoying?":* no, you see, we actually need our apps to be in control of which versions of dependencies are being used across multiple libraries, so that things like lit-element and haunted can coexist in the same app while sharing a single copy of lit-html (or different versions if you want, the point is that the app can control and configure this)

## how haunted must change to get with the program

[take a peek at haunted's current surface area](https://unpkg.com/haunted@4.4.0/)

1. **the bundle has got to go**

    - in the future, it's the default for packages to present individual es modules, and not to bundle them up
    - es module users usually don't want a bundle, they only want to load what's necessary
    - it's up to each consumer to decide whether or not they want to use bundling in their application (just like minification)
    - consider how [lit-element](https://unpkg.com/lit-element@2.1.0/lit-element.js) is used: like haunted, it also uses lit-html as a dependency, however lit-element imports lit-html with a bare-specifier, expecting you to resolve this with an import map or a bundler

    **proposed changes to haunted:**  
    haunted's build should emit individual modules, and a new main module should be elected (probably such that `import "haunted"` will resolve to `"haunted/component.js"`).  
    perhaps a legacy bundle could be included.

2. **remove the hard-baked lit-html dependency in `lit.js`**

    - currently, haunted [imports lit-html in lit.js](https://github.com/matthewp/haunted/blob/master/src/lit.js#L1) directly from unpkg
    - this is a problem, because it locks haunted to a specific version of lit-html
      - that's not cool, because it makes haunted the "dictator king" of the lit-html version being used in your entire app — and if you don't obey, you risk loading multiple copies of lit-html
      - in response, apps will need to configure their own import map to, across the app, resolve "lit-html" to the *exact same* unpkg link as haunted
      - apps are also forced to load lit-html from unpkg, they can't load it locally or from another cloud vendor like jsdelivr
    - if we don't remove the hard-baked dependency, the recommended practice will end up being to use `import {haunted} from "haunted/core.js"` and have to configure your own `component` function, even if you're just using lit-html like default

    **proposed changes to haunted:**  
    we should remove `lit.js`, and replace all of haunted's internal references to it with a bare-specifier import, like `import {render} from "lit-html"`, and expect the user to resolve the lit-html dependency with an import map or their own bundling step

some of these ideas were discussed with @matthewp and @stevenvachon in pull request #81 (making haunted agnostic with respect to lit-html/lighterhtml) which is at least somewhat related

let me know what you think :wave:

**Comment by matthewp** [comment]
@chase-moskal Thanks for posting this! Glad to continue the discussion.

The reason `src/lit.js` uses unpkg is so that the tests can be run in the browser without bundling. I'm fully in favor of non-bundling solutions which is why I made Haunted work this way. It doesn't depend on any future feature like import maps. It works today.

I understand what you're wanting but am a little confused because I think it already exists. Haunted comes with a build, `haunted/index.js` that uses the bare specifier of `lit-html`. So it will use the version you selected to download.  I don't think this is any different from how lit-element works.

Can you explain what it is about `haunted/index.js` that doesn't meet your needs?

**Comment by stevenvachon** [comment]
I'm not familiar with import maps, but if lit-html were removed from this project's source code, could it not then be imported into tests?:
```js
import haunted from 'haunted';
import {html, render} from 'lit-html';

haunted.config({html, render});

it('works', () => {});
```

**Comment by chase-moskal** [comment]
@matthewp 

> Haunted comes with a build, haunted/index.js that uses the bare specifier of lit-html

oh hey that's great, i wasn't aware that the main module for haunted *is* using bare specifiers -- i was just confused
- excellent! i suppose concern number 2 from the original post is not so much a problem :tada:
- i think i was confused because [core.js](https://github.com/matthewp/haunted/blob/master/src/core.js#L3) references to `lit.js` instead of `lit-html` -- however though i suppose there is a build step which replaces this or something like that

even still, i think it would be best to remove `lit.js` entirely, and refactor the tests something like @stevenvachon is indicating, if even to simply to avoid confusion

as for the bundle/modules:  
i think we all agree that we want to expose individual es modules, but, there might be two ways to do it

1. individual modules are the new default
    - `import "haunted"` gives you `component.js`
    - `import "haunted/bundle"` is an alternative pathway to the bundle 
2. the bundle remains the default
    - `import "haunted"` gives you the bundle
    - `import "haunted/component.js"` becomes the recommended pathway

i favor option 1, because i see bundling in the same boat as minification -- package authors should be leaving those concerns to the package users -- however i see no reason not to continue including the bundle via `haunted/bundle`

what do you think?

**Comment by matthewp** [comment]
@stevenvachon No, I don't want to require configuration to use Haunted.  I personally do not enjoy frameworks / tools where you have to install a dozen plugins and configure them together to do anything useful. Haunted should remain lit-html by default.

There may be some confusion in your example though, Haunted doesn't depend on the `html` function. It only needs `render`. This is why it works fine with [hyperHTML](https://codepen.io/WebReflection/pen/pxXrdy?editors=0010). It doesn't require that you be a lit-html user.  

Now, I don't think people not using lit-html should get the added bytes that they don't need. So that's what this discussion has always been about. The solution is to create a bundle that does not include lit-html.

**Comment by matthewp** [comment]
@chase-moskal I think I agree with you. I think we can change it so that the main and module fields points to `src/index.js`. Then we can create this new `haunted/component` (or whatever we decide to call it) that doesn't have the lit-html dependency and you can use that to configure Haunted to use lighterhtml or hyperhtml instead.

At this point I think I need to get to work and implement a version of what we've been discussing and submit a PR for you to review. I'll try to do that some time this week if possible.

**Comment by chase-moskal** [comment]
@matthewp 

> @stevenvachon No, I don't want to require configuration to use Haunted. I personally do not enjoy frameworks / tools where you have to install a dozen plugins and configure them together to do anything useful. Haunted should remain lit-html by default.

i worry you've misread what @stevenvachon is trying to say, and i do agree with him

- `lit.js` should be removed from haunted's source code, and haunted's tests should simply import `lit-html` like in his example
- this is not related to lit-html being default and ready out-of-the-box -- we agree with keeping things that way :+1:

> Now, I don't think people not using lit-html should get the added bytes that they don't need. So that's what this discussion has always been about. The solution is to create a bundle that does not include lit-html.

i'm just not really sure why you want to provide bundles at all?

- it appears to me that the emerging best practice is to leave bundling and minification to the applications -- library authors should simply expose individual es modules
- my understanding is that it's just redundant for packages to be individually minified and bundled when applications are already going to bundle and minify all of those packages?
- for example, we don't see modern packages like [lit-element](https://unpkg.com/lit-element@2.2.0/) including any bundles

anyways, neither the lit.js concern nor the bundle concern are really dealbreakers, they are just ergonomic nice-to-have's that i'd like to sell you on ;)

the most important thing that we all agree on, is that we need to expose individual es modules as a prerequisite to #81 in order to use haunted without loading lit-html -- since haunted isn't using any transpiler like typescript or babel, i think this should be implemented with a simple build step that literally copies every file from `src/` straight into the package root -- i want to stress that i think including a new bundle for every use-case is a bad pattern, and that the modules should be exposed directly unbundled

> At this point I think I need to get to work and implement a version of what we've been discussing and submit a PR for you to review.

i enjoy discussing best practices and the finer points of javascript package ergonomics, so i'm open to further discussion and i look forward to participating in that review

cheers friend! :wave:

**Comment by matthewp** [comment]
> this is not related to lit-html being default and ready out-of-the-box -- we agree with keeping things that way

I don't follow, how? If I remove lit.js then Haunted has no dependency on the `render` function that it needs.

> i'm just not really sure why you want to provide bundles at all?

I want it to run directly in the browser with no bundler required on the part of the end user. That's why there is the `haunted.js` and `web.js` bundles. Once I change the `src/` to use the bare specifier for lit-html then the src/ folder won't be loaded in the browser. So the bundles are necessary.

> cheers friend! 👋

Cheers to you! Thanks for being willing to participate here, I appreciate it a lot.

**Comment by chase-moskal** [comment]
@matthewp 

> I don't follow, how? If I remove lit.js then Haunted has no dependency on the render function that it needs.

haunted doesn't need `lit.js` to exist in order to import `lit-html` in any case -- neither for haunted's source code nor the tests -- its removal won't have any meaningful effect except to clean up the codebase

separately from the lit.js situation, i think haunted's `component.js` should become the main module (instead of the bundle) -- `component.js` should be importing haunted internals and also `lit-html`, and then exporting the component function that is configured for lit-html

> I want it to run directly in the browser with no bundler required on the part of the end user.

i don't want to take that away, i'd just like to see the bundle move to an entry point like `haunted/bundle.js`

consider how we use modern packages like `lit-element` -- they don't provide a bundle:

- we can use es-module-shims as a polyfill for import maps, it's great
- otherwise, users can use a bundle step
- yet otherwise, users can load from unpkg using the `?module` feature, which rewrites bare specifiers into full unpkg links (very cool feature)
- i think haunted should follow this pattern for which lit-element is setting a good example

to clarify what i mean

- we should remove `lit.js`
- we should keep the default behavior `import {component} from "haunted"` preconfigured for lit-html
- we should place the convenience bundle in a location like `import {component} from "haunted/bundle.js"`

looking now further into the current haunted bundles:
- the main bundle `index.js` uses bare-specifiers, and won't work "out-of-the-box" (without a bundler or es-module-shims)
- the `haunted.js` bundle uses unpkg links, and will work out of the box
- the `web.js` has relative links to `../lit-html/lit-html.js`, it requires that specific filepath (brittle)

i strongly suggest, let's make one bundle to rule them all: the bundle should be fully inclusive, including all of the dependencies like lit-html within the bundle -- the ultimate in convenience

i'm in support of keeping the bundle as a convenience, however keep in mind that `lit-element` has chosen not to provide that convenience

**Comment by matthewp** [comment]
> i think haunted should follow this pattern for which lit-element is setting a good example

I disagree that this is a good example. Web modules are available today. Lit-element requires you either bundle first or use a special server or special CDN. Using modules in the web should not wait on future features. So I think moving the bundle to some place like `haunted/bundle.js` kind of makes web modules a second class citizen and I don't want to do that.  But I don't see why we need to rename the bundles to achieve what you're wanting (a way to load the src code directly from your bundler).

As far as `lit.js`, it was done that way because it currently has the unpkg urls and it was more convenient to only have to update those in one file rather than in multiple. When I switch the `src/` to use the bare specifiers then indeed this file likely won't be needed any more.

**Comment by chase-moskal** [comment]
> So I think moving the bundle to some place like haunted/bundle.js kind of makes web modules a second class citizen and I don't want to do that.

i think maybe that's backwards? i think electing the es modules as the main entrypoint would indicate them as the first class citizen, and so it's my preference to see the bundle demoted to the second-class citizen position

either way, you know what? it's aesthetic — tomato tomato!

we agree on what actually matters: so long as both ways get to be citizens, then everybody will be well-served :)

### updated proposal

**respecting your preference for the bundle remaining the main entrypoint, here's an updated proposal**

the first two points are what i think we all already agree on, and the third point is something new i'd like to discuss about bundling

1. expose individual es modules in the package
2. remove `lit.js` in the sourcecode
3. *(for further discussion)* consider creating an "all-inclusive" bundle in which lit-html is included within the bundle
    - i think this would be an improvement over the existing bundles
    - it would be more optimal since it saves an http request to unpkg
    - for users who don't want lit-html *inside* the bundle:
      - first, i can't think of why this would be a valid use-case, i think an all-inclusive bundle should be simply superior (any reason to use a different bundle is better served using the es-modules)
      - however still, we could keep the currently existing bundles, eg `haunted/web.js`
      - again, the recommended pathway for these use-cases should be to import from `haunted/component.js` and then use import-maps, es-module-shims, or bundling in order to mandate how lit-html is provided (which is the only reason not to use an all-inclusive bundle, right?)
    - the decision here to make is: should we add this new all-inclusive bundle? if so, should it take the place as the main entrypoint for haunted? i'm in favor
      - otherwise we could leave the existing bundles as-is, and the new all-inclusive bundle can be a net-addition (perhaps located at `haunted/bundle.js`)

a major version bump is not necessarily required, unless we deprecate or rename the existing bundles (i'm in favor)

### some illustrative examples and discussion

if the above proposal was implemented in my favorite way, there would be two general ways to consume haunted

1. **the all-inclusive bundle**  
    `import {component} from "haunted"`  
    works out-of-the-box with no setup or extra thinking, great for just getting started
2. **es modules**  
    `import {component} from "haunted/component.js"`  
    requires bare specifier resolution, either via import maps, es-module-shims, or a bundler

again, i'd like to simply deprecate all of the older bundles, however, maybe we could rename them like these for added clarity -- and i've added some notes and thoughts about each bundle

```js
// using the default all-inclusive bundle which includes lit-html
// no bare-specifiers, no external dependencies/requests
// convenient, compact, and optimal
import {component} from "haunted"

// alternatively using a different bundle, this one imports lit-html from unpkg.
// this might be convenient for some people.
// however it has a gotchya/footgun: it's easy to mismatch unpkg versions
// amongst peer dependencies, which could cause double-loading of lit-html
// for that reason i would deprecate this and force people to use the es modules
import {component} from "haunted/bundle-unpkg.js"

// this bundle has bare specifiers to import lit-html.
// again, es module usage should be preferable for this use-case
import {component} from "haunted/bundle-bare.js"

// i really think we should just remove the bundle with the local filepaths
// usage of brittle local filepaths should be considered a deprecated antipattern
import {component} from "haunted/bundle-local.js"
```

we could keep the existing bundles (in which case i'd like them renamed for clarity), but i vastly prefer simply deprecating and removing the older bundles altogether because i think they are footguns compared to the proposed all-inclusive bundle

### after we're done the above work, then we can move onto #81 

once we've reached consensus on and implement the above ideas, then we should continue to implement #81 so that we can use haunted in an agnostic way with different templating libraries, like `lighterhtml`:

```js
// this import has no dependencies, does not load lit-html
import {haunted} from "haunted/core.js"

import {render} from "lighterhtml"

// configure haunted component with lighterhtml
const {component} = haunted({
  render: (what, where) => render(where, () => what)
})

export {component}
```

### tell me what you think

@matthewp, @stevenvachon, and any other observers — please weigh in with your thoughts!

even if we disagree about all of my bundle-oriented ideas with proposal point 3, it's not that important (ultimately i don't even like using bundles hahah), so i think so long as we all agree on the first two points in the proposal, we'll all be happy enough in the end — the rest is just gravy ;)

cheers! :wave:

**Comment by chase-moskal** [comment]
## plot twist!

woah! @matthewp, @stevenvachon — [the new pika cdn](https://www.pika.dev/cdn) — check this out!

pika just launched this new cdn, and it's really, *really* cool!
- you ask it for any npm package whose main entrypoint is es modules
- it will deliver to you an all-inclusive bundle
- look at how it takes these es-module projects, and gives us lovely bundles
  - [lit-element](https://cdn.pika.dev/lit-element)
  - [preact](https://cdn.pika.dev/preact)
  - [redux](https://cdn.pika.dev/redux)
  - [@google/model-viewer](http://cdn.pika.dev/@google/model-viewer)
  - there are 60,000+ more npm projects doing it this way
  - haunted is one package that does not work :(

this is exactly what we've needed, and i believe this is obviously the future of the javascript ecosystem

modules should be exposing es module as their main module entrypoint, and all of the responsibility for bundling, minification, optimization, and convenience concerns, should be left to dedicated tooling like this new pika cdn

**@matthewp — given this new development, i strongly suggest that we engineer haunted to fit into this new ecosystem that pika and others are spearheading** — make the es modules the main entrypoint, forget doing any bundling yourself, and let emerging tools like pika handle this

with the rise of pika, i think it's hard to argue that compatibility with this modern ecosystem is not a worthwhile thing for haunted

if you still disagree (hard to disagree with 60k of the most cutting edge packages on npm, isn't it?), then i'd like us to proceed with my previous post's proposal and discussion ;)

let me know if this sways you at all, or are you really that cold? hah, kidding, stay lighthearted friend, cheers :wave:

**Comment by chase-moskal** [comment]
## i'm all-in on this

you know, this new pika development is so revelatory, and i'm so confident in the advice to make es modules the main entrypoint for all packages moving into the future, that i hereby retract my proposal for haunted to include funky bundles or to allow a bundle to take the main entrypoint position

my advice is now that haunted needs to become *properly* pika compatible, and i'm no longer interested my proposal from the previous post — do what you will, but my heart and soul is definitely set on this emerging ecosystem, and moving forward, i won't consider using *any package* that hasn't been outfitted correctly to be used in this way, whether that outfitting requires a fork or not — and in the coming years, i'm sure i won't be alone with that attitude

keep in mind, this isn't about being "pika-compatible", this isn't about pika per se — pika is simply implementing these established best practices that i've been trying to evangelize earlier in this thread, but now they're forcing packages to formally acknowledge whether they are on board or not, by labeling them compatible or not

one more thing — haunted is technically pika compatible as it provides a `module` field in the package.json — however this is currently an ugly and awkward hack, because now pika is bundling the code that is already a bundle, and plus the bundles don't work for #81 agnostic... so we should just get with the program on this topic and expose es modules properly like all the other cool packages

that's my two cents, i hope i can entice you down this yellow brick road towards a harmonious javascript ecosystem where packages can be lean and minimalistic, and still provide everything everybody wants

i'm no longer interested in my previous proposal, i just want this done the modern cool way or not at all

cheers! :wave:

**Comment by matthewp** [comment]
@chase-moskal I appreciate your enthusiasm 😀 . I hope to have time to work on this in the next few days. I'll definitely design it with pika in mind.

**Comment by matthewp** [comment]
So if we get rid of the `index.js` build, how will that impact typescript? There is the `index.d.ts` file which I think describes index.js. If the new `main` is `src/haunted.js` does the typescript file need to move to the `src/` folder and become `src/haunted.d.ts`?

**Comment by chase-moskal** [comment]
@matthewp 

for each module, say `component.ts`, typescript should be configured to generate the following

- `component.js`
- `component.d.ts`
- `component.js.map`

i think the haunted package.json should have the `main` and `module` fields set to `component.js`

the package.json `types` field should be set to `component.d.ts`

when a user `import {component} from "haunted"`, typescript will refer to `component.js` and `component.d.ts` thanks to the package.json directives

typescript is also smart enough to use the relevant `.d.ts` for other modules too, eg  
`import {haunted} from "haunted/core.js"` should automatically use `core.d.ts`

i prefer names like "component" and "core" over "haunted" — it makes sense to me that haunted might contain a "component" or a "core", but it's redundant for haunted to contain a "haunted"

for small packages which have zero dependencies, i sometimes opt to use [barrelsby](https://github.com/bencoveney/barrelsby) to automatically generate an `index.ts` file, and make that the main module and set package.json types to `index.d.ts` — this is sort of like an "unbundled" bundle, which can be a convenience, and works well with tree-shaking, however i think this would be a bad practice for haunted because of the dependencies like `lit-html` (it would be too simple for people to accidentally and unknowingly import `configureComponent` through the index file and accidentally bring `lit-html` along for the ride)

**Comment by matthewp** [comment]
@chase-moskal We're not using TypeScript in this project. The `index.d.ts` file is written by hand. But since I'm going to get rid of the `index.js` bundle I'm wondering where the typings should be moved to. I'm guessing `src/haunted.d.ts` but I'm inexperienced in this.

**Comment by chase-moskal** [comment]
> We're not using TypeScript in this project. The index.d.ts file is written by hand.

oh yeah, right

i think you just place `.d.ts` files alongside each module, the same as if you had used typescript to generate them

however, if you are planning to provide and maintain typescript definition files, i strongly, strongly, *strongly* suggest that you actually use typescript — otherwise you have no means of verifying whether or not the code *actually* matches the type definitions that you are shipping — so your typings are likely to be error-prone, and as a typescript user, i might rather have no typings at all than allowing the possibility that they could be incorrect — i think typescript should be entirely embraced, or completely discarded

converting the haunted codebase to typescript sounds like it would be a big scary project, but it's really not a big deal — somebody already well-versed in typescript could enhance this codebase before lunch

**Comment by matthewp** [comment]
Yeah, I'm sure its not hard. I'm not a TypeScript user myself. If we had more regular maintainers and they all wanted to convert then I'd be open to doing that.

**Comment by matthewp** [comment]
# Notes to self

## Required

* package.json `main` and `module` fields point to `src/` folder.
* `haunted/core` - Includes only component model + hooks, no lit-html integration.
* Keep `haunted.js` bundle for web usage.
* TypeScript should still work.

## Open questions

* `haunted/core.js` should point to `src/haunted/core.js` I guess?
  * Should there be a "core bundle"? Maybe not since `src/haunted/core.js` should be loadable in a browser anyways.

**Comment by chase-moskal** [comment]
> package.json main and module fields point to src/ folder.

we've previously agreed that the package main/module fields should instead point to a `dist/` directory

in haunted's case, we're just dumping compiled output into the root directory, it's a tradeoff which calls for more stringent maintenance of the `.gitignore`, but gives us prettier looking package paths (eg `"haunted/core.js"` instead of `"haunted/dist/core.js"`) -- that's cool

of course, the reason for not pointing to `src/`, is because the sourcecode can be a moving target and be refactored, without any breaking changes to the compiled output (eg, no breaking changes for a typescript conversion in src)

so i think the relevant package.json bit should look like

```json
{
  "type": "module",
  "main": "component.js",
  "module": "component.js"
}
```

> `haunted/core.js` should point to `src/haunted/core.js` I guess?

given the above considerations, the following should hold true

```js
import {component} from "haunted"
  // resolves to "haunted/component.js" thanks to main/module fields
  // lit-html will come along for this ride, via bare specifier

import {Container} from "haunted/core.js"
  // directly gets "haunted/core.js", real path in final published package
  // no lit-html
```

> Should there be a "core bundle"?

nah, just make deprecated equivalents to the existing bundles available in the same locations, remove their recommendation from the readme, and drop a line about how users looking for a bundle should grab a fresh one from the pika cdn otherwise use rollup or their next favorite tool

> - [ ] Update tests to use the haunted.js bundle.

why do you want the tests to use a bundle? is this just an artifact of a workflow that you are familiar with? why not use es modules in your tests?

**Comment by matthewp** [comment]
The `haunted.js` bundle *is* an ES module. There are no non-modules in this codebase. The haunted.js bundle is because bare module specifiers do not work in the browser. The haunted.js bundle instead uses `unpkg.com` to load lit-html.

Good point on `src/`, could possibly rename to `lib/`.

**Comment by chase-moskal** [comment]
> The haunted.js bundle is an ES module. There are no non-modules in this codebase. The haunted.js bundle is because bare module specifiers do not work in the browser. The haunted.js bundle instead uses unpkg.com to load lit-html.

sure sure, but that's the deprecated way -- what happens with the old bundle isn't interesting, what's interesting is how haunted will interplay with the modern ecosystem -- for that we need our individualized module outputs, and that's all that matters to me (and is an architectural prerequisite for #81)

we'll tell haunted users that they should use a pika bundle, or do their own bundling

the old `haunted.js` bundle being deprecated, doesn't need to be mentioned or promoted in the readme

> Good point on src/, could possibly rename to lib/.

- some people like sourcecode in `src`, and distribution code in `dist`
- some people prefer to use the term `lib` instead of `dist`
- the problem, is that some people like to use `lib` instead of `src`, and then emit to `dist`
- so when we see somebody use `lib`, we can't be sure if that's sourcecode, or if that's consumable output

if `lib` is where the sourcecode lives -- then of course we should never point to `lib`, because we want to be able to refactor the sourcecode safely without affecting the emitted output

if you want to call the consumable output `lib`, then there's this minor ambiguity problem

so it seems like the hip thing to do is to keep it simple and unambiguous -- `src` for sourcecode, `dist` for what gets distributed -- and we just never use `lib`, in order to avoid the confusion altogether

haunted should take one of two options here
1. `haunted/src/` outputs to `haunted/dist/`
2. `haunted/src/` outputs to `haunted/`

personally i think number 1 is cleaner and simpler from a maintenance perspective, but number 2 provides more attractive-looking import paths for your consumers (at the expense of a grossly maintained `.gitignore`) -- it's a toss-up, but it's subjective

**Comment by chase-moskal** [comment]
i wanted to reiterate -- the tests shouldn't be using a bundle

each test suite should import the modules that it is testing, like any modern code

:wave: Chase

**Comment by matthewp** [comment]
To clarify, `haunted.js` is not deprecated and is still going to be mentioned in the readme. I disagree that using browser modules is not "modern" enough. I don't want to require that people use special tools like pika to use Haunted. They should be able import it directly from the browser; no tools required.

> i wanted to reiterate -- the tests shouldn't be using a bundle
> each test suite should import the modules that it is testing, like any modern code

They can't, the browser doesn't support bare module specifiers.

**Comment by stevenvachon** [comment]
I like being able to `npm install` and work locally. I don't know, but I think using URLs in `import` statements will complicate that. I also use webpack/parcel and babel for everything.

**Comment by matthewp** [comment]
@stevenvachon If you use a bundler (or any other tool like pika) the package.json `main` and `module` fields will work. The haunted.js is only for browser usage.

**Comment by chase-moskal** [comment]
@matthewp 

> To clarify, haunted.js is not deprecated and is still going to be mentioned in the readme.

okay, just bear in mind that one downside, is a minor aesthetic wart for haunted moving into the future
- people of the future will prefer `import {component} from "haunted/component.js"`
- instead of simply `import {component} from "haunted"`
- avoiding the bundle is more optimal, better tree-shaking

> I disagree that using browser modules is not "modern" enough.

i'm not sure if this statement is confusing because it's a misunderstanding — of course in every case we both agree that we want people to be using es modules — i just don't want the main module to be the bundle — it's the *bundles* that aren't "modern" enough

> They can't, the browser doesn't support bare module specifiers.

that's what the es-module-shims are for ;)

well, if you're going to keep the `haunted.js` bundle around, i suppose you might as well continue to use it in the tests..

i think my temperament is to introduce people to modern workflows, and your temperament is to respect and preserve the workflows they are already familiar with — i respect your decision

i'll focus my review on the side of haunted that matters to the future workflows, and stop complaining about the bundle taking the main/module throne ;)

@stevenvachon 

> I like being able to npm install and work locally. I don't know, but I think using URLs in import statements will complicate that. I also use webpack/parcel and babel for everything.

after the work related to this discussion is merged, insofar as we're on the same page, @stevenvachon, for your workflow you're describing, you'll want to use haunted this way

```js
import {component} from "haunted/component.js"
```

this module will have a bare-specifier dependency on `lit-html`, which your bundler can resolve from your local `node_modules` — no URLs

if you want an optimized tree-shaken bundle, you'd then be able to grab one from pika for `haunted/component.js`

@matthewp 

so barring our disagreement about which file takes the top main/module spot — i think we're on the same page for the rest of it — let me know if there's anything about what i'm saying that makes you queasy and indicates otherwise ;)

cheers!
:wave: Chase

**Comment by matthewp** [comment]
>  i just don't want the main module to be the bundle — it's the bundles that aren't "modern" enough

The main module is not the haunted.js bundle. Maybe that's the confusion. the package.json `main` will point to the unbundled source. If you're using a bundler, pika, or anything else that parses the package.json, you'll never be using any bundles. The only people who will use a bundle is people who use the browser `<script type="module">` and manually type in the url `https://unpkg.com/haunted/hautned.js`.  Does that make you feel better?

**Comment by chase-moskal** [comment]
@matthewp — yes it does make me feel much better!

cheers!

:wave: Chase
